src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecution.java

```diff
@@ -63,7 +69,19 @@
  * @author Kohsuke Kawaguchi
  * @author Jesse Glick
  */
-public abstract class FlowExecution implements FlowActionStorage {
+public abstract class FlowExecution implements FlowActionStorage, GraphLookupView {  // Implements GraphLookupView because FlowNode lives in another package
+
+    protected transient GraphLookupView internalGraphLookup = null;
+
+    /** Eventually this may be overridden if the FlowExecution has a better source of structural information, such as the {@link FlowNode} storage. */
+    protected synchronized GraphLookupView getInternalGraphLookup() {
```

Comment:
Still recommend this be `private` or `@Restricted` unless there is a specific override in hand now.